go/types/list.go

```diff
@@ -170,33 +173,39 @@ func (l List) IterAll(f listIterAllFunc) {
 	}
 }
 
-func (l List) copyReadAhead(out []Value, startIdx uint64) (numBytes uint64) {
-	llen := l.Len()
-	d.PanicIfFalse(startIdx < llen)
+func copyReadAhead(col Collection, blockLength uint64, startIdx uint64) (out []Value, numBytes uint64) {
+	d.PanicIfTrue(blockLength == 0)
 
-	endIdx := startIdx + uint64(len(out))
-	if endIdx > llen {
-		endIdx = llen
-	}
+	l := col.Len()
+	d.PanicIfFalse(startIdx < l)
 
-	if startIdx == endIdx {
-		return
+	endIdx := startIdx + blockLength
+	d.PanicIfFalse(endIdx <= l)
+
+	valuesPerIdx := uint64(1)
+	if col.Kind() == MapKind {
+		valuesPerIdx = 2
```

Comment:
you could make this part of the collection interface.

Reply:
eh doesn't seem worthwhile updating all those call sites. I made it a function though.